DatadogTrace/Sources/OpenTelemetry/OTelSpan.swift

```diff
@@ -0,0 +1,162 @@
+        self.queue = tracer.queue
+        self.tracer = tracer
+        self.nestedSpan = .init(
+            tracer: tracer,
+            context: .init(
+                traceID: context.traceId.toDatadog(),
+                spanID: context.spanId.toDatadog(),
+                parentSpanID: parentSpanID?.toDatadog(),
+                baggageItems: .init()
+            ),
+            operationName: name,
+            startTime: startTime,
+            tags: [:]
+        )
+    }
+
+    // swiftlint:disable unavailable_function
+    func addEvent(name: String) {
+        fatalError("Not implemented yet")
+    }
+
+    func addEvent(name: String, timestamp: Date) {
+        fatalError("Not implemented yet")
+    }
+
+    func addEvent(name: String, attributes: [String: OpenTelemetryApi.AttributeValue]) {
+        fatalError("Not implemented yet")
+    }
+
+    func addEvent(name: String, attributes: [String: OpenTelemetryApi.AttributeValue], timestamp: Date) {
+        fatalError("Not implemented yet")
+    }
+    // swiftlint:enable unavailable_function
+
+    func end() {
+        end(time: Date())
+    }
+
+    func end(time: Date) {
+        queue.sync {
+            guard isRecording else {
+                return
+            }
+            isRecording = false
+
+            // Attributes maps to tags in Datadog
+            for (key, value) in attributes {
+                switch value {
+                case .string(let value):
+                    nestedSpan.setTag(key: key, value: value)
+                case .bool(let value):
+                    nestedSpan.setTag(key: key, value: value.description)
+                case .int(let value):
+                    nestedSpan.setTag(key: key, value: value.description)
+                case .double(let value):
+                    nestedSpan.setTag(key: key, value: value.description)
+                // swiftlint:disable unavailable_function
+                case .stringArray:
+                    fatalError("Not implemented yet")
+                case .boolArray:
+                    fatalError("Not implemented yet")
+                case .intArray:
+                    fatalError("Not implemented yet")
+                case .doubleArray:
+                    fatalError("Not implemented yet")
+                case .set:
+                    fatalError("Not implemented yet")
+                // swiftlint:enable unavailable_function
+                }
+                nestedSpan.setTag(key: key, value: value.description)
+            }
+
+            // SpanKind maps to the `span.kind` tag in Datadog
+            nestedSpan.setTag(key: DatadogTagKeys.spanKind.rawValue, value: kind.rawValue)
+        }
+        nestedSpan.finish(at: time)
+    }
+
+    var description: String {
+        return "OTelSpan"
+    }
+
+    func setAttribute(key: String, value: OpenTelemetryApi.AttributeValue?) {
+        guard isRecording else {
+            return
+        }
+
+        attributes[key] = value
```

Comment:
blocking/ This can be called from any thread and `attributes` is not synchronised. We should either use `queue` or `@ReadWriteLock` on `var attributes`.

Reply:
yup, clear miss on my side.